dashboard/app/models/user.rb

```diff
@@ -217,7 +217,7 @@ class User < ActiveRecord::Base
   has_many :districts_users, class_name: 'DistrictsUsers'
   has_many :districts, through: :districts_users
 
-  has_many :authentication_options, dependent: :destroy
+  has_many :authentication_options, dependent: :destroy, autosave: true
```

Comment:
I'm not sure how I feel about adding this. Previously, without `autosave: true`, only newly created authentication options were autosaved when the user saved (see [ActiveRecord::AutosaveAssociation](https://api.rubyonrails.org/classes/ActiveRecord/AutosaveAssociation.html)).

I added autosave because when the user changes their user_type, we want their authentication options to re-validate and re-save. More specifically, when a user goes from teacher -> student, we want the teacher's authentication options to clear all plaintext email addresses.

I could do this work manually, but wanted to see how folks felt about always autosaving authentication options when a user is saved.

Reply:
This breaks the pattern we've been following where we only create/destroy `AuthenticationOption`, never update it; that said, this makes a long of sense in the case you're describing, and I'm not opposed to it.

In fact, when an account is changed to a student account we may need to also locate soft-deleted AuthenticationOptions for that user and make sure they also have cleartext emails removed.

Reply:
ah, yeah, good point. re: your question about this happening elsewhere in the system, i would think this only happens when a user changes their user type, since that's the only time we care about clearing plaintext email addresses.

maybe, instead of autosaving, we add a `before_save` that conditionally runs when the user is changing type from teacher -> student and nukes the email addresses in all auth options? something like:
```
def remove_student_cleartext_emails
  authentication_options.with_deleted.update_all(email: '') if student?
end
```

Reply:
I like that better, yeah.  Good call.